File: airbyte-workers/src/test/java/io/airbyte/workers/temporal/discover/catalog/DiscoverCatalogActivityTest.java
/*
 * Copyright (c) 2020-2024 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.temporal.discover.catalog;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.when;

import io.airbyte.api.client.AirbyteApiClient;
import io.airbyte.api.client.model.generated.Geography;
import io.airbyte.commons.features.FeatureFlags;
import io.airbyte.commons.protocol.AirbyteMessageSerDeProvider;
import io.airbyte.commons.protocol.AirbyteProtocolVersionedMigratorFactory;
import io.airbyte.commons.workers.config.WorkerConfigsProvider;
import io.airbyte.config.ActorContext;
import io.airbyte.config.Configs;
import io.airbyte.config.ConnectorJobOutput;
import io.airbyte.config.StandardDiscoverCatalogInput;
import io.airbyte.config.helpers.LogConfigs;
import io.airbyte.config.secrets.SecretsRepositoryReader;
import io.airbyte.featureflag.ConfigFileClient;
import io.airbyte.featureflag.FeatureFlagClient;
import io.airbyte.metrics.lib.MetricClient;
import io.airbyte.persistence.job.models.IntegrationLauncherConfig;
import io.airbyte.persistence.job.models.JobRunConfig;
import io.airbyte.workers.exception.WorkerException;
import io.airbyte.workers.helper.GsonPksExtractor;
import io.airbyte.workers.models.DiscoverCatalogInput;
import io.airbyte.workers.process.ProcessFactory;
import io.airbyte.workers.workload.JobOutputDocStore;
import io.airbyte.workers.workload.WorkloadIdGenerator;
import io.airbyte.workers.workload.exception.DocStoreAccessException;
import io.airbyte.workload.api.client.generated.WorkloadApi;
import java.nio.file.Path;
import java.util.Optional;
import java.util.UUID;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class DiscoverCatalogActivityTest {

  private final WorkerConfigsProvider workerConfigsProvider = mock();
  private final ProcessFactory processFactory = mock();
  private final SecretsRepositoryReader secretsRepositoryReader = mock();
  private final Path workspaceRoot = mock();
  private final Configs.WorkerEnvironment workerEnvironment = mock();
  private final LogConfigs logConfigs = mock();
  private final AirbyteApiClient airbyteApiClient = mock();
  private final String airbyteVersion = "";
  private final AirbyteMessageSerDeProvider serDeProvider = mock();
  private final AirbyteProtocolVersionedMigratorFactory migratorFactory = mock();
  private final FeatureFlags featureFlags = mock();
  private final MetricClient metricClient = mock();
  private final FeatureFlagClient featureFlagClient = mock(ConfigFileClient.class);
  private final GsonPksExtractor gsonPksExtractor = mock();
  private final WorkloadApi workloadApi = mock();
  private final WorkloadIdGenerator workloadIdGenerator = mock();
  private final JobOutputDocStore jobOutputDocStore = mock();
  private DiscoverCatalogActivityImpl discoverCatalogActivity;

  @BeforeEach
  void init() {
    discoverCatalogActivity = spy(new DiscoverCatalogActivityImpl(
        workerConfigsProvider,
        processFactory,
        secretsRepositoryReader,
        workspaceRoot,
        workerEnvironment,
        logConfigs,
        airbyteApiClient,
        airbyteVersion,
        serDeProvider,
        migratorFactory,
        featureFlags,
        metricClient,
        featureFlagClient,
        gsonPksExtractor,
        workloadApi,
        workloadIdGenerator,
        jobOutputDocStore));
  }

  @Test
  void runWithWorkload() throws DocStoreAccessException, WorkerException {
    final String jobId = "123";
    final int attemptNumber = 456;
    final UUID actorDefinitionId = UUID.randomUUID();
    final String workloadId = "789";
    final UUID workspaceId = UUID.randomUUID();
    final UUID connectionId = UUID.randomUUID();

    final DiscoverCatalogInput input = new DiscoverCatalogInput();
    input.setJobRunConfig(new JobRunConfig()
        .withJobId(jobId)
        .withAttemptId((long) attemptNumber));
    input.setDiscoverCatalogInput(new StandardDiscoverCatalogInput()
        .withActorContext(new ActorContext()
            .withWorkspaceId(workspaceId)
            .withActorDefinitionId(actorDefinitionId)));
    input.setLauncherConfig(new IntegrationLauncherConfig().withConnectionId(connectionId));

    when(workloadIdGenerator.generateDiscoverWorkloadId(actorDefinitionId, jobId, attemptNumber)).thenReturn(workloadId);
    doReturn(Geography.AUTO).when(discoverCatalogActivity).getGeography(any(), any());
    doReturn(true).when(discoverCatalogActivity).isWorkloadTerminal(any());

    final ConnectorJobOutput output = new ConnectorJobOutput().withOutputType(ConnectorJobOutput.OutputType.DISCOVER_CATALOG_ID)
        .withDiscoverCatalogId(UUID.randomUUID());
    when(jobOutputDocStore.read(workloadId)).thenReturn(Optional.of(output));

    ConnectorJobOutput actualOutput = discoverCatalogActivity.runWithWorkload(input);

    assertEquals(output, actualOutput);
  }

}

File: airbyte-workers/src/test/kotlin/io/airbyte/workers/temporal/discover/catalog/DiscoverCatalogActivityTest.kt
/*
 * Copyright (c) 2020-2024 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.workers.temporal.discover.catalog

import io.airbyte.api.client.AirbyteApiClient
import io.airbyte.api.client.generated.ConnectionApi
import io.airbyte.commons.logging.DEFAULT_LOG_FILENAME
import io.airbyte.commons.logging.LogClientManager
import io.airbyte.config.ActorContext
import io.airbyte.config.ConnectorJobOutput
import io.airbyte.config.StandardDiscoverCatalogInput
import io.airbyte.config.WorkloadPriority
import io.airbyte.featureflag.FeatureFlagClient
import io.airbyte.featureflag.TestClient
import io.airbyte.featureflag.WorkloadCheckFrequencyInSeconds
import io.airbyte.persistence.job.models.IntegrationLauncherConfig
import io.airbyte.persistence.job.models.JobRunConfig
import io.airbyte.workers.commands.DiscoverCommand
import io.airbyte.workers.models.DiscoverCatalogInput
import io.airbyte.workers.sync.WorkloadClient
import io.airbyte.workers.workload.WorkloadIdGenerator
import io.airbyte.workload.api.client.model.generated.WorkloadCreateRequest
import io.airbyte.workload.api.client.model.generated.WorkloadType
import io.mockk.CapturingSlot
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.spyk
import io.mockk.verify
import io.temporal.activity.ActivityExecutionContext
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import java.nio.file.Path
import java.util.UUID
import kotlin.time.Duration.Companion.minutes
import kotlin.time.DurationUnit

class DiscoverCatalogActivityTest {
  private val discoverCatalogSnapDuration = 15.minutes

  private val workspaceRoot: Path = Path.of("workspace-root")
  private val airbyteApiClient: AirbyteApiClient = mockk()
  private val featureFlagClient: FeatureFlagClient = spyk(TestClient())
  private val connectionApi: ConnectionApi = mockk()
  private val workloadClient: WorkloadClient = mockk()
  private val workloadIdGenerator: WorkloadIdGenerator = mockk()
  private val logClientManager: LogClientManager = mockk()
  private val executionContext: ActivityExecutionContext =
    mockk {
      every { heartbeat(null) } returns Unit
    }
  private lateinit var createReqSlot: CapturingSlot<WorkloadCreateRequest>
  private lateinit var discoverCatalogActivity: DiscoverCatalogActivityImpl
  private lateinit var discoverCommand: DiscoverCommand

  @BeforeEach
  fun init() {
    every { airbyteApiClient.connectionApi }.returns(connectionApi)
    discoverCommand =
      spyk(
        DiscoverCommand(
          workspaceRoot = workspaceRoot,
          airbyteApiClient = airbyteApiClient,
          workloadClient = workloadClient,
          workloadIdGenerator = workloadIdGenerator,
          logClientManager = logClientManager,
          discoverAutoRefreshWindowMinutes = discoverCatalogSnapDuration.toInt(DurationUnit.MINUTES),
        ),
      )
    discoverCatalogActivity =
      spyk(
        DiscoverCatalogActivityImpl(
          featureFlagClient,
          workloadClient,
          discoverCommand,
        ),
      )
    every { discoverCatalogActivity.activityContext } returns executionContext
    every { logClientManager.fullLogPath(any()) } answers { Path.of(invocation.args[0].toString(), DEFAULT_LOG_FILENAME).toString() }
    every { featureFlagClient.intVariation(WorkloadCheckFrequencyInSeconds, any()) } returns WORKLOAD_CHECK_FREQUENCY_IN_SECONDS

    createReqSlot = slot<WorkloadCreateRequest>()
    every {
      workloadClient.runWorkloadWithCancellationHeartbeat(
        capture(createReqSlot),
        WORKLOAD_CHECK_FREQUENCY_IN_SECONDS,
        executionContext,
      )
    } returns Unit
  }

  @ParameterizedTest
  @ValueSource(booleans = [true, false])
  fun runWithWorkload(runAsPartOfSync: Boolean) {
    val jobId = "123"
    val attemptNumber = 456
    val actorDefinitionId = UUID.randomUUID()
    val actorId = UUID.randomUUID()
    val workloadId = "789"
    val workspaceId = UUID.randomUUID()
    val connectionId = UUID.randomUUID()
    val input =
      DiscoverCatalogInput(
        jobRunConfig =
          JobRunConfig()
            .withJobId(jobId)
            .withAttemptId(attemptNumber.toLong()),
        discoverCatalogInput =
          StandardDiscoverCatalogInput()
            .withActorContext(
              ActorContext()
                .withWorkspaceId(workspaceId)
                .withActorDefinitionId(actorDefinitionId)
                .withActorId(actorId),
            ).withManual(!runAsPartOfSync),
        launcherConfig =
          IntegrationLauncherConfig()
            .withConnectionId(
              connectionId,
            ).withWorkspaceId(workspaceId)
            .withPriority(WorkloadPriority.DEFAULT),
      )
    if (runAsPartOfSync) {
      every {
        workloadIdGenerator.generateDiscoverWorkloadIdV2WithSnap(eq(actorId), any(), eq(discoverCatalogSnapDuration.inWholeMilliseconds))
      }.returns(workloadId)
    } else {
      every { workloadIdGenerator.generateDiscoverWorkloadId(actorDefinitionId, jobId, attemptNumber) }.returns(workloadId)
    }

    val output =
      ConnectorJobOutput()
        .withOutputType(ConnectorJobOutput.OutputType.DISCOVER_CATALOG_ID)
        .withDiscoverCatalogId(UUID.randomUUID())
    every { workloadClient.getConnectorJobOutput(workloadId, any()) } returns output

    val result = discoverCatalogActivity.runWithWorkload(input)

    verify { workloadClient.runWorkloadWithCancellationHeartbeat(any(), WORKLOAD_CHECK_FREQUENCY_IN_SECONDS, executionContext) }
    assertEquals(workloadId, createReqSlot.captured.workloadId)
    assertEquals(WorkloadType.DISCOVER, createReqSlot.captured.type)
    assertEquals(ConnectorJobOutput.OutputType.DISCOVER_CATALOG_ID, output.outputType)
    assertEquals(output, result)
  }

  companion object {
    private const val WORKLOAD_CHECK_FREQUENCY_IN_SECONDS = 10
  }
}
